sharding-orchestration/sharding-orchestration-core/src/main/java/io/shardingsphere/orchestration/internal/registry/config/listener/SchemaChangedListener.java

```diff
@@ -0,0 +1,138 @@
+    
+    private final Collection<String> existedSchemaNames = new LinkedList<>();
+    
+    public SchemaChangedListener(final String name, final RegistryCenter regCenter, final Collection<String> shardingSchemaNames) {
+        super(regCenter, new ConfigurationNode(name).getSchemaPath());
+        configurationService = new ConfigurationService(name, regCenter);
+        configurationNode = new ConfigurationNode(name);
+        existedSchemaNames.addAll(shardingSchemaNames);
+    }
+    
+    @Override
+    protected ShardingOrchestrationEvent createShardingOrchestrationEvent(final DataChangedEvent event) {
+        String shardingSchemaName = configurationNode.getSchemaName(event.getKey());
+        if (Strings.isNullOrEmpty(shardingSchemaName) || !isValidNodeChangedEvent(shardingSchemaName, event.getKey())) {
+            return new IgnoredChangedEvent();
+        }
+        if (ChangedType.UPDATED == event.getChangedType()) {
+            return createUpdateChangedEvent(shardingSchemaName, event);
+        }
+        if (ChangedType.DELETED == event.getChangedType()) {
+            return createDeleteChangedEvent(shardingSchemaName);
+        }
+        return new IgnoredChangedEvent();
+    }
+    
+    private boolean isValidNodeChangedEvent(final String shardingSchemaName, final String nodeFullPath) {
+        return configurationNode.getDataSourcePath(shardingSchemaName).equals(nodeFullPath) || configurationNode.getRulePath(shardingSchemaName).equals(nodeFullPath);
+    }
+    
+    private ShardingOrchestrationEvent createUpdateChangedEvent(final String shardingSchemaName, final DataChangedEvent event) {
+        if (existedSchemaNames.contains(shardingSchemaName)) {
+            return createChangedEventForExistedSchema(event, shardingSchemaName);
+        }
+        return createChangedEventForNewSchema(shardingSchemaName);
+    }
+    
+    private ShardingOrchestrationEvent createChangedEventForExistedSchema(final DataChangedEvent event, final String shardingSchemaName) {
+        if (event.getKey().equals(configurationNode.getDataSourcePath(shardingSchemaName))) {
+            return createDataSourceChangedEvent(shardingSchemaName, event);
+        }
+        return createRuleChangedEvent(shardingSchemaName, event);
+    }
+    
+    private DataSourceChangedEvent createDataSourceChangedEvent(final String shardingSchemaName, final DataChangedEvent event) {
+        return new DataSourceChangedEvent(shardingSchemaName, ConfigurationYamlConverter.loadDataSourceConfigurations(event.getValue()));
+    }
+    
+    private ShardingOrchestrationEvent createRuleChangedEvent(final String shardingSchemaName, final DataChangedEvent event) {
+        return isShardingRule(event) ? createShardingRuleChangedEvent(shardingSchemaName, event.getValue()) : createMasterSlaveRuleChangedEvent(shardingSchemaName, event.getValue());
+    }
+    
+    private boolean isShardingRule(final DataChangedEvent event) {
```

Comment:
consider about move this logic to a common place?

Reply:
It is only used here, what do your mean of common place?

Reply:
you can search `"tables:\n"`, some places use it